File: sdk/csharp/src/SeocheonSdk.cs
using Seocheon.Sdk.Constants;
using Seocheon.Sdk.Errors;
using Seocheon.Sdk.Infrastructure;
using Seocheon.Sdk.Infrastructure.Signing;
using Seocheon.Sdk.Internal.Tx;
using Seocheon.Sdk.Modules;

namespace Seocheon.Sdk;

/// <summary>
/// Main Seocheon SDK client. Entry point for all blockchain operations.
/// </summary>
public sealed class SeocheonSdk : IDisposable
{
    private readonly SdkConfig _config;
    private readonly IChainClient _client;
    private readonly ISigningService _signer;
    private bool _connected;

    /// <summary>Activity submission and query.</summary>
    public ActivityModule Activity { get; }

    /// <summary>Epoch and window information.</summary>
    public EpochModule Epoch { get; }

    /// <summary>Node information and search.</summary>
    public NodeModule Node { get; }

    /// <summary>Rewards query and withdrawal.</summary>
    public RewardsModule Rewards { get; }

    /// <summary>Balance, transfer, block, and TX queries.</summary>
    public CosmosModule Cosmos { get; }

    /// <summary>
    /// Creates a new SDK instance with the given configuration.
    /// </summary>
    /// <param name="config">SDK configuration.</param>
    /// <param name="chainClient">Optional custom chain client (for testing).</param>
    /// <param name="signingService">Optional custom signing service (for testing).</param>
    public SeocheonSdk(SdkConfig config, IChainClient? chainClient = null, ISigningService? signingService = null)
    {
        config.Validate();
        _config = config;

        _client = chainClient ?? new HttpChainClient(config.Chain.RpcEndpoint, config.Chain.GrpcEndpoint);
        _signer = signingService ?? CreateSigner(config.Signing);

        var pipelineConfig = new PipelineConfig
        {
            ChainId = config.Chain.ChainId,
            GasPrice = ParseGasPrice(config.Chain.GasPrice),
            ConfirmTimeout = TimeSpan.FromMilliseconds(config.Tx.ConfirmTimeoutMs),
            PollInterval = TimeSpan.FromMilliseconds(config.Tx.ConfirmPollIntervalMs)
        };

        Activity = new ActivityModule(_client, _signer, pipelineConfig);
        Epoch = new EpochModule(_client, _signer);
        Node = new NodeModule(_client, _signer);
        Rewards = new RewardsModule(_client, _signer, pipelineConfig);
        Cosmos = new CosmosModule(_client, _signer, pipelineConfig);
    }

    /// <summary>
    /// Connects to the blockchain.
    /// </summary>
    public async Task ConnectAsync(CancellationToken ct = default)
    {
        if (_connected)
            return;

        await _client.ConnectAsync(ct);
        _connected = true;
    }

    /// <summary>
    /// Disconnects from the blockchain.
    /// </summary>
    public async Task DisconnectAsync()
    {
        if (!_connected)
            return;

        await _client.DisconnectAsync();
        _connected = false;
    }

    /// <summary>
    /// Returns whether the SDK is connected to the chain.
    /// </summary>
    public bool IsConnected => _connected;

    /// <summary>
    /// Returns the current SDK configuration.
    /// </summary>
    public SdkConfig Config => _config;

    /// <summary>
    /// Returns the signer's address.
    /// </summary>
    public string GetAddress() => _signer.GetAddress();

    /// <summary>
    /// Disposes the SDK and releases resources.
    /// </summary>
    public void Dispose()
    {
        DisconnectAsync().GetAwaiter().GetResult();
        if (_client is IDisposable disposable)
            disposable.Dispose();
    }

    private static ISigningService CreateSigner(SigningConfig config)
    {
        return config.Mode switch
        {
            SigningMode.Direct => new DirectService(config.Mnemonic),
            SigningMode.Keystore => new KeystoreService(
                config.KeystorePath,
                Environment.GetEnvironmentVariable(config.PassphraseEnv)
                    ?? throw SdkErrors.InvalidConfig($"Environment variable '{config.PassphraseEnv}' not set")
            ),
            SigningMode.Vault => new VaultService(config.VaultEndpoint, config.KeyName),
            _ => throw SdkErrors.InvalidConfig($"Unknown signing mode: {config.Mode}")
        };
    }

    private static ulong ParseGasPrice(string gasPriceStr)
    {
        // Parse "250uppyeo" → 250
        var numStr = new string(gasPriceStr.TakeWhile(char.IsDigit).ToArray());
        return ulong.TryParse(numStr, out var price) ? price : ChainConstants.DefaultGasPrice;
    }
}

File: sdk/kotlin/src/main/kotlin/com/seocheon/sdk/SeocheonSdk.kt
package com.seocheon.sdk

import com.seocheon.sdk.errors.SdkError
import com.seocheon.sdk.infrastructure.ChainClient
import com.seocheon.sdk.infrastructure.OkHttpChainClient
import com.seocheon.sdk.infrastructure.signing.DirectService
import com.seocheon.sdk.infrastructure.signing.KeystoreService
import com.seocheon.sdk.infrastructure.signing.SigningService
import com.seocheon.sdk.infrastructure.signing.VaultService
import com.seocheon.sdk.internal.tx.PipelineConfig
import com.seocheon.sdk.modules.*

/**
 * Main Seocheon SDK client.
 * Provides access to 5 modules: activity, node, epoch, rewards, cosmos.
 */
class SeocheonSdk private constructor(
    private val config: SDKConfig,
    private val chainClient: ChainClient,
    private val signer: SigningService,
) {
    val activity: ActivityModule
    val node: NodeModule
    val epoch: EpochModule
    val rewards: RewardsModule
    val cosmos: CosmosModule

    private var connected = false

    init {
        val pipelineConfig = PipelineConfig(
            chainId = config.chain.chainId,
            gasPrice = parseGasPrice(config.chain.gasPrice),
            confirmTimeoutMs = config.tx.confirmTimeoutMs,
            pollIntervalMs = config.tx.confirmPollIntervalMs,
        )

        activity = ActivityModule(chainClient, signer, pipelineConfig)
        node = NodeModule(chainClient, signer, pipelineConfig)
        epoch = EpochModule(chainClient)
        rewards = RewardsModule(chainClient, signer, pipelineConfig)
        cosmos = CosmosModule(chainClient, signer, pipelineConfig)
    }

    /**
     * Connects to the blockchain.
     */
    suspend fun connect() {
        chainClient.connect()
        connected = true
    }

    /**
     * Disconnects from the blockchain.
     */
    suspend fun disconnect() {
        chainClient.disconnect()
        connected = false
    }

    /**
     * Returns true if the SDK is connected.
     */
    fun isConnected(): Boolean = connected && chainClient.isConnected()

    /**
     * Returns the SDK configuration.
     */
    fun getConfig(): SDKConfig = config

    /**
     * Returns the signer's address.
     */
    fun getAddress(): String = signer.getAddress()

    companion object {
        /**
         * Creates a new SeocheonSdk instance.
         * Validates config and creates the appropriate signing service.
         */
        suspend fun create(config: SDKConfig): SeocheonSdk {
            config.validate()

            val chainClient = OkHttpChainClient(config.chain)
            val signer = createSigner(config.signing)

            return SeocheonSdk(config, chainClient, signer)
        }

        /**
         * Creates a SeocheonSdk with a custom chain client (for testing).
         */
        fun createWithClient(config: SDKConfig, chainClient: ChainClient, signer: SigningService): SeocheonSdk {
            return SeocheonSdk(config, chainClient, signer)
        }

        private suspend fun createSigner(signing: SigningConfig): SigningService = when (signing.mode) {
            SigningMode.DIRECT -> DirectService(signing.mnemonic)
            SigningMode.KEYSTORE -> KeystoreService.create(signing.keystorePath, signing.passphraseEnv)
            SigningMode.VAULT -> VaultService.create(signing.vaultEndpoint, signing.keyName)
        }

        private fun parseGasPrice(gasPrice: String): Long {
            val numStr = gasPrice.replace(Regex("[^0-9]"), "")
            return numStr.toLongOrNull() ?: 250L
        }
    }
}
